perception/autoware_ground_segmentation/src/scan_ground_filter/grid.hpp

```diff
@@ -17,93 +17,481 @@
 
 #include <autoware/universe_utils/geometry/geometry.hpp>
 #include <autoware/universe_utils/math/normalization.hpp>
+#include <autoware/universe_utils/system/time_keeper.hpp>
 
+#include <glob.h>
+
+#include <algorithm>
 #include <cmath>
+#include <memory>
+#include <utility>
+#include <vector>
+
+namespace
+{
+
+float pseudoArcTan2(const float y, const float x)
+{
+  // lightweight arc tangent
+
+  // avoid divide-by-zero
+  if (y == 0.0f) {
+    if (x >= 0.0f) return 0.0f;
+    return M_PIf;
+  }
+  if (x == 0.0f) {
+    if (y >= 0.0f) return M_PI_2f;
+    return -M_PI_2f;
+  }
+
+  const float x_abs = std::abs(x);
+  const float y_abs = std::abs(y);
+
+  // divide to 8 zones
+  constexpr float M_2PIf = 2.0f * M_PIf;
+  constexpr float M_3PI_2f = 3.0f * M_PI_2f;
+  if (x_abs > y_abs) {
+    const float ratio = y_abs / x_abs;
+    const float angle = ratio * M_PI_4f;
+    if (y >= 0.0f) {
+      if (x >= 0.0f) return angle;  // 1st zone
+      return M_PIf - angle;         // 2nd zone
+    } else {
+      if (x >= 0.0f) return M_2PIf - angle;  // 4th zone
+      return M_PIf + angle;                  // 3rd zone
+    }
+  } else {
+    const float ratio = x_abs / y_abs;
+    const float angle = ratio * M_PI_4f;
+    if (y >= 0.0f) {
+      if (x >= 0.0f) return M_PI_2f - angle;  // 1st zone
+      return M_PI_2f + angle;                 // 2nd zone
+    } else {
+      if (x >= 0.0f) return M_3PI_2f + angle;  // 4th zone
+      return M_3PI_2f - angle;                 // 3rd zone
+    }
+  }
+}
+
+float pseudoTan(const float theta)
+{
+  // lightweight tangent
+
+  // normalize the angle, range of [-pi/2, pi/2]
+  float normalized_theta = theta;
+  while (normalized_theta > M_PI_2f) {
+    normalized_theta -= M_PIf;
+  }
+  while (normalized_theta < -M_PI_2f) {
+    normalized_theta += M_PIf;
+  }
+
+  // avoid divide-by-zero
+  if (normalized_theta == 0.0f) return 0.0f;
+
+  if (std::abs(normalized_theta) <= 1.0f) {
+    return normalized_theta / M_PI_4f;
+  }
+  return std::copysign(M_PI_4f / (M_PI_2f - std::abs(normalized_theta)), normalized_theta);
+}
+}  // namespace
 
 namespace autoware::ground_segmentation
 {
+using autoware::universe_utils::ScopedTimeTrack;
+
+struct Point
+{
+  size_t index;
+  float distance;
+  float height;
+};
 
-class ScanGroundGrid
+// Concentric Zone Model (CZM) based polar grid
+class Cell
 {
 public:
-  ScanGroundGrid() = default;
-  ~ScanGroundGrid() = default;
+  // list of point indices
+  std::vector<Point> point_list_;  // point index and distance
+
+  // method to check if the cell is empty
+  inline bool isEmpty() const { return point_list_.empty(); }
+  inline int getPointNum() const { return point_list_.size(); }
```

Comment:
IMO, it is better to change to make sure safer type return.
```suggestion
  inline int getPointNum() const { return static_cast<int>(point_list_.size()); }
```